crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2036,20 +2065,38 @@ impl AuthorityPerEpochStore {
 
         let mut batch = self.db_batch();
 
-        sequenced_transactions.extend(
-            self.load_deferred_transactions_for_consensus_round(&mut batch, commit_round)?
-                .into_iter(),
-        );
+        // Pre-process transactions to find the most recent randomness round included in the commit.
+        let mut last_randomness_round_written = self.last_randomness_round_written()?;
+        for tx in sequenced_transactions.iter() {
+            if let SequencedConsensusTransactionKind::System(tx) = &tx.0.transaction {
+                if let TransactionKind::RandomnessStateUpdate(rsu) =
+                    tx.data().intent_message().value.kind()
+                {
+                    last_randomness_round_written =
+                        std::cmp::max(last_randomness_round_written, rsu.randomness_round);
+                }
+            }
+        }
 
-        // TODO: This is a no-op until we start using random round transactions
-        let placeholder_random_round = u64::MAX - 1;
-        sequenced_transactions.extend(
-            self.load_deferred_transactions_for_randomness_round(
+        // Load transactions deferred from prevous commits.
+        let deferred_tx: Vec<VerifiedSequencedConsensusTransaction> = self
+            .load_deferred_transactions_for_consensus_round(&mut batch, commit_round)?
+            .into_iter()
+            .chain(self.load_deferred_transactions_for_randomness_round(
                 &mut batch,
-                placeholder_random_round,
-            )?
-            .into_iter(),
-        );
+                last_randomness_round_written,
+            )?)
+            .collect();
+        let deferred_tx_digests: HashSet<_> = deferred_tx
+            .iter()
+            .map(|tx| match tx.0.transaction.key() {
+                SequencedConsensusTransactionKey::External(
+                    ConsensusTransactionKey::Certificate(digest),
+                ) => digest,
+                _ => panic!("deferred transaction was not a user certificate: {tx:?}"),
+            })
+            .collect();
+        sequenced_transactions.extend(deferred_tx.into_iter());
```

Comment:
this is getting quite subtle so at the very least we need some comments/asserts:
- we have loaded every deferred tx that is due in the new random round
- we want every tx to be sequenced using exactly the random round it was scheduled for
- this means that
   - a) we must sequence the RandomnessStateUpdate before the readers we just loaded (this is ensured by the `reorder` call, but that is not obvious from reading the code)
   - b) we depend on the assumption that there is only one RSU per narwhal commit. We should probably assert for this condition when we scan the transactions above to find the new randomness round.